### PR TITLE
Store ping timestamp per host separately.

### PR DESCRIPTION
**Pull request by ankilosaurus**
Problem:
----------
Host table keeps inventory of all the hosts registered to Teletraan.
1. It keeps timestamp of all the registered hosts -> used by AgentJanitor to cleanup unreachable or dead hosts.
2. It also keep mapping between host and group. (For example: helloworlddummyservice-server-dev-0a01def3 <> CMP, helloworlddummyservice-server-dev-0a01def3 <> mcrouter-prod-sidecar, ... etc.) -> used in UI/API to populate hosts which are part of the specific environment.

Due to this timestamp needs to be updated per host <-> group mapping for a given host instead of just per host.

Fix:
----
1. Extend existing hosts_and_agents table to keep per host data (specifically timestamp). 
With this change, 
- last_update field of hosts table will not be updated on every ping. Instead last_update in hosts_and_agents table will be updated on every ping once per host id.
- host <-> group mapping will be inserted once for new host and then after for any addition/deletion of host <-> group mapping.
This potentially reduces inserts by 7 folds per host per ping.
(Note: fields from hosts table are not removed for backward compatibility)

2. Improved on AgentJanitor logic by removing group level processing as that loop iterate through all the groups in the system to find unreachable hosts -> which I think is a processing waste. 
AgentJanitor will continue to go through all the stale hosts in the system with added asg info in hosts_and_agents table so as to pull host launch latency threshold config from ASG for a given stale host.

--
Other fields - state, can_retire - can also be moved to host_and_agents table as those are also per host. However, as that requires substantial changes at API level and also client side,  those changes are not included in the scope.

Tests:
---
1. Verify host_name, host_id, ip, asg, create_date, last_update, agent_version of a host in hosts_and_agents table.
2. Verify last_update field is not updated in hosts table on every ping if there is no change in host <-> group mapping.
3. Verify that new deploy groups are updated in hosts table.
4. Verify that stale deploy groups are removed from hosts table.
2. Terminate host part of ASG and verify it's cleaned up.
2. Remove deployd from host part of ASG and verify it's marked unreachable.
3. Terminate standalone hosts and verify it's cleaned up.
4. Remove deployd from standalone host and verify it's marked unreachable.

### CODE REVIEW

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java [hunk 1]
```diff
@@ -144,17 +144,50 @@ public DeployBean load(String deployId) throws Exception {
...
+                LOG.warn("Remove host {} to group {} mapping", hostName, recordedGroup);
                 this.hostDAO.removeHostFromGroup(hostId, recordedGroup);
             }
         }
+    }
 
-
+    void updateHostStatus(String hostId, String hostName, String hostIp, String agentVersion, String asg, Set<String> groups) throws Exception {
+        HostAgentBean hostAgentBean = hostAgentDAO.getHostById(hostId);
+        if (hostAgentBean == null) {
```
Comment (disparter) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:172]: I prefer to split this function into 2 more to improve code readness. 
This isn't a requirement we can discuss this if want.

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java [hunk 2]
```diff
@@ -49,18 +50,23 @@ public AgentJanitor(ServiceContext serviceContext, int minStaleHostThreshold,
...
             Collection<String> resultIds = rodimusManager.getTerminatedHosts(Collections.singletonList(hostId));
             if (resultIds.isEmpty()) {
                 terminatedHosts.remove(hostId);
             }
         }
+        return terminatedHosts;
+    }
+
+    private void processStaleHosts(Collection<String> staleHostIds, boolean isMarkedUnreachable) throws Exception {
+        if (staleHostIds.isEmpty()) {
```
Comment (disparter) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java:65]: This collection can be nullable? if yes I guess you could check using Collections

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java [hunk 3]
```diff
@@ -87,113 +93,60 @@ private void processStaleHosts(Collection<String> staleHostIds, boolean isMarked
     }
 
     private Long getInstanceLaunchGracePeriod(String clusterName) throws Exception {
-        Long launchGracePeriod = rodimusManager.getClusterInstanceLaunchGracePeriod(clusterName);
+        Long launchGracePeriod = (clusterName != null) ? rodimusManager.getClusterInstanceLaunchGracePeriod(clusterName) : null;
```
Comment (disparter) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java:96]: You don't need parenthesis.

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java [hunk 4]
```diff
@@ -87,113 +93,60 @@ private void processStaleHosts(Collection<String> staleHostIds, boolean isMarked
...
+        long current_time = System.currentTimeMillis();
         // If host fails to ping for longer than min stale threshold,
         // either mark them as UNREACHABLE, or remove if confirmed with source of truth
         long minThreshold = current_time - minStaleHostThreshold;
-        List<HostBean> minStaleHosts = hostDAO.getStaleHosts(minThreshold);
+        List<HostAgentBean> minStaleHosts = hostAgentDAO.getStaleHosts(minThreshold);
         Set<String> minStaleHostIds = new HashSet<>();
-        for (HostBean host : minStaleHosts) {
-            minStaleHostIds.add(host.getHost_id());
+        for (HostAgentBean hostAgentBean : minStaleHosts) {
```
Comment (disparter) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java:108]: What about using lambdas in these 2 fours?
Maybe can improve readability.

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java [hunk 5]
```diff
@@ -87,113 +93,60 @@ private void processStaleHosts(Collection<String> staleHostIds, boolean isMarked
...
         Set<String> staleHostIds = new HashSet<>();
-        for (HostBean host : maxStaleHosts) {
+        
+        for (HostAgentBean hostAgentBean : maxStaleHosts) {
+            Long launchGracePeriod = getInstanceLaunchGracePeriod(hostAgentBean.getAuto_scaling_group());
+            HostBean host = hostDAO.getHostsByHostId(hostAgentBean.getHost_id()).get(0);
             if (host.getState() == HostState.PROVISIONED) {
-                if (current_time - host.getLast_update() >= maxLaunchLatencyThreshold) {
-                    staleHostIds.add(host.getHost_id());
+                if (current_time - hostAgentBean.getLast_update() >= launchGracePeriod) {
```
Comment (disparter) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java:128]: Maybe you can merge the two ifs.
Also can extract the ifelse logic as well.
Not a requirement, just an option

Comment (ankilosaurus) [reply deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java:131]: thought of not touching this part on the first go but I think it makes sense.

Comment (ankilosaurus) [reply deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java:131]: diff updated.

**Review by disparter — APPROVED** [review]

File: deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBHostAgentDAOImpl.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 Pinterest, Inc.
+ * Copyright 2016 Pinterest, Inc.
```
[the remark this answers is not in the document]
Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBHostAgentDAOImpl.java:2]: actually, it was a typo. this file was introduced last year.

**Review by yongwen — APPROVED** [review]

**Pull request merged**